### PR TITLE
PIX: Account for alignment when computing member offset

### PR DESCRIPTION
**Pull request by jeffnn**
This has been a long-standing issue for PIX. Certain structs (with e.g., embedded arrays of structs) don't pay attention to their alignment requirements.

### CODE REVIEW

**Comment by Keenuts** [comment]
Hi! Thanks for the PR. Don't own the DXIL part, so cannot approve, but tried it, and I see no difference between in the generated DXIL for this shader taken from the test between this PR and HEAD (a0b6fc1028b673228e61dba413606de5d4b4c458), is that expected?

```hlsl
struct EmbeddedStruct
{
    uint32_t TwoDArray[2][2];
};

struct smallPayload
{
    uint32_t OneInt;
    EmbeddedStruct embeddedStruct;
    uint64_t bigOne;
};

[numthreads(1, 1, 1)]
void main()
{
    smallPayload p;
    p.OneInt = -137;
    p.embeddedStruct.TwoDArray[0][0] = 252;
    p.embeddedStruct.TwoDArray[0][1] = 253;
    p.embeddedStruct.TwoDArray[1][0] = 254;
    p.embeddedStruct.TwoDArray[1][1] = 255;
    p.bigOne = 123456789;

    DispatchMesh(2, 1, 1, p);
}
```

cmd: `./build/bin/dxc ./repro.hlsl -T as_6_5 -E main -Zi`

**Comment by jeffnn** [comment]
> Hi! Thanks for the PR. Don't own the DXIL part, so cannot approve, but tried it, and I see no difference between in the generated DXIL for this shader taken from the test between this PR and HEAD ([a0b6fc1](https://github.com/microsoft/DirectXShaderCompiler/commit/a0b6fc1028b673228e61dba413606de5d4b4c458)), is that expected?
> 
> ```hlsl
> struct EmbeddedStruct
> {
>     uint32_t TwoDArray[2][2];
> };
> 
> struct smallPayload
> {
>     uint32_t OneInt;
>     EmbeddedStruct embeddedStruct;
>     uint64_t bigOne;
> };
> 
> [numthreads(1, 1, 1)]
> void main()
> {
>     smallPayload p;
>     p.OneInt = -137;
>     p.embeddedStruct.TwoDArray[0][0] = 252;
>     p.embeddedStruct.TwoDArray[0][1] = 253;
>     p.embeddedStruct.TwoDArray[1][0] = 254;
>     p.embeddedStruct.TwoDArray[1][1] = 255;
>     p.bigOne = 123456789;
> 
>     DispatchMesh(2, 1, 1, p);
> }
> ```
> 
> cmd: `./build/bin/dxc ./repro.hlsl -T as_6_5 -E main -Zi`

Hi. This isn't expected to produce any difference in generated code. It's related to how debug information is processed, specifically to how that debug information reflects placement (according to the various structure packing rules) for embedded arrays like in the example.

File: lib/DxilDia/DxilDiaSymbolManager.cpp [hunk 1]
```diff
@@ -1140,13 +1143,23 @@ void dxil_dia::hlsl_symbols::SymbolManagerInit::TypeInfo::Embed(const TypeInfo &
   for (const auto &E : TI.GetLayout()) {
     m_Layout.emplace_back(E);
   }
+  uint64_t alignmentInBytes = TI.GetAlignmentInBytes();
+  if (alignmentInBytes != 0 && 
+      m_dwCurrentSizeInBytes % alignmentInBytes != 0) {
+    m_dwCurrentSizeInBytes += alignmentInBytes - (m_dwCurrentSizeInBytes % alignmentInBytes);
```
Comment (adam-yang) [inline lib/DxilDia/DxilDiaSymbolManager.cpp:1149]: Isn't this basically llvm::RoundUpToAlignment? Same with in TypeInfo::AddBasicType.

Comment (jeffnn) [reply lib/DxilDia/DxilDiaSymbolManager.cpp]: So it is! Didn't know about that. Changed to use it.

**Review by adam-yang — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**